django/db/models/fields/files.py

```diff
@@ -313,12 +314,16 @@ def generate_filename(self, instance, filename):
         Until the storage layer, all file paths are expected to be Unix style
         (with forward slashes).
         """
-        filename = validate_file_name(filename)
+        # Use base names for absolute paths, to prevent potential
+        # directory-traversal.
+        if os.path.isabs(filename):
+            filename = os.path.basename(filename)
```

Comment:
I have doubts about these two lines, using absolute paths didn't work before 0b79eb36915d178aef5c6a7bbce71b1e76d376d3.

`model_fields.test_filefield.FileFieldTests.test_save_without_name` crashes without them with:
```
django.core.exceptions.SuspiciousFileOperation:
Detected path traversal attempt in '/tmp/django_1m9we578/tmplongmnm7.txt'
```
in Django 3.2 (before 0b79eb36915d178aef5c6a7bbce71b1e76d376d3) it crashes with:
```
django.core.exceptions.SuspiciousFileOperation:
The joined path (/tmp/django_glv73wg6/tmptqw34_9u.txt) is located outside of the base path component (/django/tests)
```

IMO we should remove them and change `test_filefield_generate_filename_absolute_path` to catch `django.core.exceptions.SuspiciousFileOperation: Detected path traversal attempt in ...`.

Reply:
I added the 2nd commit to show how it looks like without absolute path support. IMO, much better.

Reply:
I tend to agree. Using basename here doesn't give us much. If one uses a `upload_to` callable they can modify it to not return a absolute path (same situation as before, because otherwise it would fail at the storage layer) and if one simply has a string it will continue to fail like before. In theory we could take the basename in the latter case, but that would be a change in behavior.

Reply:
It's not necessary for my use case, as I'm using a custom `upload_to` callable which only takes the extension from the filename. But I tend to lean towards having the basename in place. It wouldn't break anything, only simplify working with string `upload_to` values.

Without it, this is not possible (with string `upload_to`):
`model.file_field = File(open('/tmp/image.png', 'rb'))`
And providing the `name` argument for `File` doesn't seem like a good solution, as it should contain the full path to the file.

Reply:
@kukosk It never worked, so it is not a regression per se. We shouldn't add extra cleanups/new features to a patch that it's intended to backport.

Reply:
@felixxm Yes, that makes total sense 👍